File: mobile/lib/core/capabilities/capabilities.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../shared/providers/core_providers.dart';

/// What this account may see and do, as the server works it out.
///
/// ---- Why the app is told rather than deciding ---------------------------
///
/// The alternative is a switch in here: `if (plan == 'hospital') showTabs()`.
/// That puts the product's shape in the client, where it ships on a release
/// cycle measured in app-store reviews and where two phones disagree the moment
/// one is a version behind. A practice upgrading its plan would keep seeing the
/// old navigation until every doctor updated the app.
///
/// So the server answers *what is available* and the app answers *where to put
/// it*. A capability the server stops sending disappears on the next load, on
/// every device at once.
///
/// ---- And this is not a security boundary -------------------------------
///
/// Nothing here is trusted. The routes are guarded server-side; this exists so
/// the app does not offer a button that would come back 403, which is a
/// courtesy to the person holding the phone rather than a control on them.
class Capabilities {
  const Capabilities({
    required this.practiceType,
    required this.specialty,
    required this.plan,
    required this.practice,
    required this.effective,
    required this.role,
    required this.isOwner,
    required this.resolved,
    this.hasDietician = false,
    this.permissions = const <String>{},
    this.ui,
    this.department,
  });

  /// What kind of organisation. Null for a practice nobody has classified,
  /// which is every one created before types existed.
  final String? practiceType;
  final String? specialty;
  final String? plan;

  /// What the organisation has.
  final Set<String> practice;

  /// What this person may use — the smaller of the two, and the one to build
  /// navigation from.
  final Set<String> effective;

  final String? role;
  final bool isOwner;

  /// What this person may do, resolved.
  ///
  /// The server sends the role's preset where nobody has customised the grant,
  /// rather than the empty array it stores — two readings of "empty" is how a
  /// screen comes to hide every button from the person who owns the practice.
  final Set<String> permissions;

  /// Whether anybody at this practice writes diet plans.
  ///
  /// Not a capability — a capability is what the product offers, this is who
  /// the practice employs. It arrives in the same response because the
  /// navigation needs it and that is the request the navigation already makes.
  final bool hasDietician;

  /// Has an answer actually arrived?
  ///
  /// False means "not known yet", which is a different thing from "nothing is
  /// available" and has to be, or an empty set would read as a locked-down
  /// account. Every check below leans on this.
  final bool resolved;

  /// Which components this person's home screen is made of, and in what order.
  ///
  /// Null before an answer arrives, and null for a patient — who has no
  /// clinician home screen rather than an empty one. See [DashboardConfig].
  final DashboardConfig? ui;

  /// Which part of the practice they work in. Null for a solo clinic, and for
  /// anybody nobody has assigned — both ordinary, neither a failure.
  final DepartmentRef? department;

  /// The permissive default, used before the first response arrives.
  ///
  /// Everything on, because the server is the thing that refuses and a screen
  /// hidden while a request is in flight is a screen that flickers away under
  /// somebody's thumb. A button that turns out not to work is a worse
  /// experience than a slow one; a navigation bar that rearranges itself twice
  /// on every cold start is worse than both.
  ///
  /// Which is also the rule the server follows — see the note on absence in
  /// `services/capabilities.js`. Both ends treat "unknown" as "do not narrow",
  /// and they have to agree or one of them is hiding what the other permits.
  static const unknown = Capabilities(
    practiceType: null,
    specialty: null,
    plan: null,
    practice: <String>{},
    effective: <String>{},
    role: null,
    isOwner: false,
    resolved: false,
  );

  bool has(String capability) => !resolved || effective.contains(capability);

  /// Whether this person holds a permission.
  ///
  /// Permissive before the answer arrives, exactly as [has] is and for the same
  /// reason: a button that appears a moment late is better than one that
  /// vanishes under somebody's thumb, and the server refuses what it should
  /// either way.
  ///
  /// A caller with no membership — a patient, or an account the backfill has
  /// not reached — holds nothing here, and every screen using it is a
  /// clinician's screen.
  bool can(String permission) => !resolved || permissions.contains(permission);

  /// True when the practice has it and this person does not — the case worth
  /// saying something different about, because one is a sale and the other is
  /// a conversation with whoever runs the practice.
  bool withheld(String capability) =>
      resolved && practice.contains(capability) && !effective.contains(capability);

  factory Capabilities.fromJson(Map<String, dynamic> json) {
    final membership = json['membership'] as Map<String, dynamic>?;
    return Capabilities(
      resolved: true,
      practiceType: json['practiceType'] as String?,
      specialty: json['specialty'] as String?,
      plan: json['plan'] as String?,
      practice: ((json['practice'] as List?) ?? const [])
          .map((e) => e.toString())
          .toSet(),
      effective: ((json['effective'] as List?) ?? const [])
          .map((e) => e.toString())
          .toSet(),
      role: membership?['role'] as String?,
      isOwner: membership?['isOwner'] == true,
      hasDietician: json['hasDietician'] == true,
      permissions: ((membership?['permissions'] as List?) ?? const [])
          .map((e) => e.toString())
          .toSet(),
      ui: DashboardConfig.fromJson(json['ui'] as Map<String, dynamic>?),
      department: DepartmentRef.fromJson(
        membership?['department'] as Map<String, dynamic>?,
      ),
    );
  }
}

/// A department, named.
///
/// The key is what code reasons about and the name is what a reader sees —
/// already resolved to this account's language by the server, because the
/// fallback ladder (their language, then English, then the key) is one rule
/// and belongs in one place.
class DepartmentRef {
  const DepartmentRef({required this.key, required this.name});

  final String key;
  final String name;

  static DepartmentRef? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    final key = json['key'] as String?;
    if (key == null) return null;
    return DepartmentRef(key: key, name: json['name'] as String? ?? key);
  }
}

/// What this person's home screen is made of.
///
/// ---- Identifiers, and the app decides everything else ------------------
///
/// The server sends names from `services/uiConfig.js` — never layout, never
/// styling, never data. How a component looks, what it fetches and how it
/// behaves are this app's business; whether it is on the screen at all is the
/// server's. A response that carried the other half would be a backend able to
/// draw anything it liked on a clinician's phone.
///
/// Which also means an unknown name is dropped here, not rendered as an error.
/// The server drops what it does not recognise and so does the registry — an
/// app a version behind meets a component that did not exist when it shipped,
/// and the right response is a slightly shorter dashboard rather than a red
/// screen in front of a patient.
class DashboardConfig {
  const DashboardConfig({
    required this.widgets,
    required this.quickActions,
    this.department,
  });

  final List<String> widgets;
  final List<String> quickActions;

  /// The department key this arrangement came from, or null for the general
  /// clinical set.
  final String? department;

  static DashboardConfig? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    return DashboardConfig(
      widgets: ((json['widgets'] as List?) ?? const [])
          .map((e) => e.toString())
          .toList(growable: false),
      quickActions: ((json['quickActions'] as List?) ?? const [])
          .map((e) => e.toString())
          .toList(growable: false),
      department: json['department'] as String?,
    );
  }
}

/// What a person may do, as opposed to what the product offers.
///
/// Mirrors `PERMISSIONS` in models/Membership.js. Written out rather than used
/// as bare strings so a typo is a compile error instead of a feature that is
/// quietly always off — a drifted name resolves to a string the server never
/// sends, `can()` returns false, and the button is gone for everybody, which
/// looks exactly like a decision somebody made on purpose.
abstract final class Perm {
  static const viewPatient = 'VIEW_PATIENT';
  static const editRecord = 'EDIT_RECORD';
  static const prescribe = 'PRESCRIBE';
  static const manageStaff = 'MANAGE_STAFF';
  static const manageDepartment = 'MANAGE_DEPARTMENT';
  static const viewAudit = 'VIEW_AUDIT';
  static const shareRecords = 'SHARE_RECORDS';
}

/// The capability names the app checks. Mirrors `services/capabilities.js`.
///
/// Written out for the same reason [Perm] is, and it is the same failure: a
/// drifted name is a capability the server never sends, and the feature is
/// quietly off everywhere.
abstract final class Cap {
  static const prescription = 'PRESCRIPTION';
  static const labOrder = 'LAB_ORDER';
  static const labResult = 'LAB_RESULT';
  static const department = 'DEPARTMENT';
  static const multiLocation = 'MULTI_LOCATION';
  static const aiAssistant = 'AI_ASSISTANT';
  static const advancedAnalytics = 'ADVANCED_ANALYTICS';
  static const advancedReports = 'ADVANCED_REPORTS';
  static const reportExport = 'REPORT_EXPORT';
  static const departmentAnalytics = 'DEPARTMENT_ANALYTICS';
  static const staffAnalytics = 'STAFF_ANALYTICS';
  static const scheduledReports = 'SCHEDULED_REPORTS';
}

/// Fetched once per sign-in and cached.
///
/// Not `autoDispose`: this is read by the navigation, so disposing it when the
/// last screen using it goes away means re-fetching on the next tab change.
final capabilitiesProvider = FutureProvider<Capabilities>((ref) async {
  final json = await ref.watch(apiClientProvider).getJson('/me/capabilities');
  return Capabilities.fromJson(json);
});

/// The answer without the loading state, for a widget that only needs to know
/// whether to draw something.
///
/// Falls back to [Capabilities.unknown] while loading and on failure. A failed
/// capability fetch must not empty the app — the routes still refuse what they
/// should, and a doctor whose network blipped should see their practice rather
/// than a stripped-down version of it.
final capabilitySetProvider = Provider<Capabilities>((ref) {
  return ref.watch(capabilitiesProvider).valueOrNull ?? Capabilities.unknown;
});

File: backend/src/services/capabilities.js
import { PRACTICE_TYPE, PLAN } from '../models/Practice.js';
import { PERMISSIONS } from '../models/Membership.js';
import { ROLES } from '../models/User.js';

/**
 * What this practice can do, and what this person in it can do.
 *
 * ---- Why this exists rather than a flag ---------------------------------
 *
 * The alternative was `premiumFeatures: true`, and the reason not to is that it
 * answers the wrong question. "Is this practice premium" is a billing fact.
 * "May this screen order a lab" is a product fact, and they are not the same
 * question — a diagnostic centre on the cheapest plan orders labs, and a
 * diabetes clinic on the dearest one has no use for a department.
 *
 * A boolean forces every screen to re-derive the second from the first, so the
 * derivation lives in twenty places and disagrees with itself in three of them.
 * This file is that derivation, once.
 *
 * ---- Four inputs, one answer --------------------------------------------
 *
 *   practice type   what kind of organisation this is at all
 *   plan            what the customer has paid for
 *   grants          what an operator has turned on for this practice by hand
 *   permissions     what THIS person may do inside it
 *
 * The first three decide what the practice HAS. The fourth decides what the
 * person in front of the screen may USE. Both are needed: a receptionist at a
 * hospital cannot prescribe, and neither can a doctor at a diagnostic centre —
 * for entirely different reasons, and a system that models only one of them
 * gets the other wrong.
 *
 * ---- Absence permits ----------------------------------------------------
 *
 * The rule the whole tenant migration rests on, applied again here. A practice
 * with no type is not a practice with no capabilities; it is a practice nobody
 * has classified yet, and every one that exists today is one. A resolver that
 * treated null as "nothing" would take prescriptions away from the clinic
 * seeing patients this morning, on the deploy that added a field.
 *
 * So: unknown type contributes everything, unknown plan contributes everything,
 * and narrowing begins the moment somebody says what this practice is.
 */

/**
 * The product surfaces that can be switched on or off.
 *
 * Named for what somebody does, not for the screen it lives on. `PRESCRIPTION`
 * survives the prescription screen being rewritten; `PRESCRIPTION_TAB` does not.
 */
export const CAPABILITIES = Object.freeze({
  PRESCRIPTION: 'PRESCRIPTION',
  LAB_ORDER: 'LAB_ORDER',
  LAB_RESULT: 'LAB_RESULT',
  DEPARTMENT: 'DEPARTMENT',
  MULTI_LOCATION: 'MULTI_LOCATION',
  AI_ASSISTANT: 'AI_ASSISTANT',
  ADVANCED_ANALYTICS: 'ADVANCED_ANALYTICS',
  ADVANCED_REPORTS: 'ADVANCED_REPORTS',
  REPORT_EXPORT: 'REPORT_EXPORT',
  DEPARTMENT_ANALYTICS: 'DEPARTMENT_ANALYTICS',
  STAFF_ANALYTICS: 'STAFF_ANALYTICS',
  SCHEDULED_REPORTS: 'SCHEDULED_REPORTS',
});

const C = CAPABILITIES;

/** Every capability there is. The permissive answer, and the starting point. */
const ALL = Object.freeze(Object.values(CAPABILITIES));

/**
 * What each kind of organisation can have at all.
 *
 * This is a ceiling, not a grant — a hospital on the cheapest plan does not get
 * advanced analytics because hospitals are listed as capable of it. The plan
 * table below is the other ceiling, and a capability has to clear both.
 *
 * A diagnostic centre is the interesting row: it orders and reports labs and
 * does not prescribe, which is not a plan decision or a permission decision. It
 * is what the organisation is licensed to do.
 */
const BY_TYPE = Object.freeze({
  [PRACTICE_TYPE.CLINIC]: [
    C.PRESCRIPTION,
    C.LAB_ORDER,
    C.LAB_RESULT,
    C.AI_ASSISTANT,
    C.ADVANCED_ANALYTICS,
    C.ADVANCED_REPORTS,
    C.REPORT_EXPORT,
    C.STAFF_ANALYTICS,
  ],

  // A specialty centre is a clinic that is organised into sub-specialties.
  [PRACTICE_TYPE.SPECIALTY_CENTRE]: [
    C.PRESCRIPTION,
    C.LAB_ORDER,
    C.LAB_RESULT,
    C.DEPARTMENT,
    C.AI_ASSISTANT,
    C.ADVANCED_ANALYTICS,
    C.ADVANCED_REPORTS,
    C.REPORT_EXPORT,
    C.DEPARTMENT_ANALYTICS,
    C.STAFF_ANALYTICS,
  ],

  // Labs in, results out, and no prescribing. The one row where a capability
  // is withheld for a reason that is neither money nor seniority.
  [PRACTICE_TYPE.DIAGNOSTIC_CENTRE]: [
    C.LAB_ORDER,
    C.LAB_RESULT,
    C.MULTI_LOCATION,
    C.ADVANCED_ANALYTICS,
    C.ADVANCED_REPORTS,
    C.REPORT_EXPORT,
    C.STAFF_ANALYTICS,
    C.SCHEDULED_REPORTS,
  ],

  [PRACTICE_TYPE.POLYCLINIC]: [
    C.PRESCRIPTION,
    C.LAB_ORDER,
    C.LAB_RESULT,
    C.DEPARTMENT,
    C.MULTI_LOCATION,
    C.AI_ASSISTANT,
    C.ADVANCED_ANALYTICS,
    C.ADVANCED_REPORTS,
    C.REPORT_EXPORT,
    C.DEPARTMENT_ANALYTICS,
    C.STAFF_ANALYTICS,
  ],

  [PRACTICE_TYPE.HOSPITAL]: ALL,
  [PRACTICE_TYPE.HEALTHCARE_GROUP]: ALL,
});

/**
 * What each plan pays for.
 *
 * A plan is a name and a set of capabilities, and nothing else. It carries no
 * price — that lives in the Razorpay dashboard, which is where somebody can
 * change it without a deploy — and no numeric limit, which is per practice in
 * `Practice.limits` so that a customer who negotiates an extra location does
 * not need a plan invented for them.
 *
 * TRIAL is deliberately the widest. Somebody deciding whether to buy should be
 * looking at the product, not at a version of it with the interesting parts
 * removed — and a trial that cannot demonstrate what is being sold converts
 * nobody.
 */
const BY_PLAN = Object.freeze({
  [PLAN.TRIAL]: ALL,

  // The core clinical loop, and the basic numbers that come with it.
  [PLAN.ESSENTIAL]: [C.PRESCRIPTION, C.LAB_ORDER, C.LAB_RESULT, C.AI_ASSISTANT],

  [PLAN.PROFESSIONAL]: [
    C.PRESCRIPTION,
    C.LAB_ORDER,
    C.LAB_RESULT,
    C.DEPARTMENT,
    C.MULTI_LOCATION,
    C.AI_ASSISTANT,
    C.ADVANCED_ANALYTICS,
    C.ADVANCED_REPORTS,
    C.REPORT_EXPORT,
    C.DEPARTMENT_ANALYTICS,
    C.STAFF_ANALYTICS,
  ],

  [PLAN.ENTERPRISE]: ALL,
});

/**
 * Which permission a capability needs before this person may use it.
 *
 * Not every capability needs one. `AI_ASSISTANT` is available to anybody who
 * may see a patient, and what it may say about that patient is decided by the
 * patient scoping rather than by a permission — see [ai/scope.js].
 *
 * A capability with no entry here is available to every member of a practice
 * that has it.
 */
const NEEDS_PERMISSION = Object.freeze({
  [C.PRESCRIPTION]: PERMISSIONS.PRESCRIBE,
  [C.LAB_ORDER]: PERMISSIONS.EDIT_RECORD,
  [C.DEPARTMENT]: PERMISSIONS.MANAGE_DEPARTMENT,
  [C.DEPARTMENT_ANALYTICS]: PERMISSIONS.MANAGE_DEPARTMENT,
  [C.STAFF_ANALYTICS]: PERMISSIONS.MANAGE_STAFF,
  [C.SCHEDULED_REPORTS]: PERMISSIONS.MANAGE_STAFF,
});

/**
 * And which capabilities a role never gets, whatever the grant says.
 *
 * A short list on purpose. Most of "what may this person do" belongs in the
 * permission grant on their membership row, which is per-person and editable;
 * a role exclusion is per-role and needs a deploy to change. Only the ones that
 * would be incoherent belong here.
 */
const ROLE_EXCLUDES = Object.freeze({
  // A dietician writes diet plans. Whatever a practice has bought, and whatever
  // an over-generous grant says, that is not prescribing.
  [ROLES.DIETICIAN]: [C.PRESCRIPTION],
  // The desk books, registers and takes payment. It does not order clinical
  // investigations, and MANAGE_STAFF on a practice manager must not start
  // meaning it.
  [ROLES.STAFF]: [C.PRESCRIPTION, C.LAB_ORDER],

  /*
   * Assists on the record and does not sign.
   *
   * The line between assisting and practising, and the one exclusion that
   * defines this role — the permission grant is the desk's, so without this
   * an assistant with PRESCRIBE granted by hand would be able to sign.
   */
  [ROLES.DOCTOR_ASSISTANT]: [C.PRESCRIPTION],

  /*
   * The laboratory roles, and the reason they had to exist.
   *
   * Filed as `staff` they lost LAB_ORDER — deliberately, because a
   * receptionist must not order clinical investigations — and the result was
   * a bench that could read a critical result and could not begin the work
   * that produced it. Neither of these excludes LAB_ORDER or LAB_RESULT;
   * both exclude PRESCRIPTION, because reporting a result is not treating
   * anybody.
   */
  [ROLES.LAB_MANAGER]: [C.PRESCRIPTION],
  [ROLES.LAB_TECHNICIAN]: [C.PRESCRIPTION],

  /*
   * Administers the practice and touches no clinical work at all.
   *
   * The longest list here, and the only one that is a statement about the job
   * rather than about seniority. A practice manager rosters people, runs
   * departments and handles billing; they do not prescribe, order a test or
   * read a result — and MANAGE_STAFF, which they need, must never start to
   * mean any of those.
   */
  [ROLES.PRACTICE_MANAGER]: [C.PRESCRIPTION, C.LAB_ORDER, C.LAB_RESULT, C.AI_ASSISTANT],
});

/** A capability list for a key, or `null` meaning "unknown, do not narrow". */
function ceiling(table, key) {
  if (!key) return null;
  const row = table[key];
  return row ? [...row] : null;
}

/**
 * What this practice has, before anybody is standing in front of it.
 *
 * `grants` is the per-practice override an operator sets — the Enterprise
 * escape hatch, and the way to turn something on for one customer without
 * inventing a plan. It only ever adds: an operator switching a capability on
 * for a practice whose type cannot have it is a mistake worth refusing, and the
 * type ceiling is the thing that refuses it.
 */
export function capabilitiesOfPractice(practice) {
  const byType = ceiling(BY_TYPE, practice?.practiceType);
  const byPlan = ceiling(BY_PLAN, practice?.plan);

  // Unknown on either axis contributes nothing to the narrowing. This is the
  // whole safety argument and it should stay the first thing this function does
  // with its inputs.
  let out = ALL.filter(
    (c) => (byType === null || byType.includes(c)) && (byPlan === null || byPlan.includes(c)),
  );

  const grants = practice?.capabilities;
  if (Array.isArray(grants) && grants.length) {
    // Added, then re-checked against the type ceiling. A grant cannot make a
    // diagnostic centre prescribe.
    const added = grants.filter(
      (c) => ALL.includes(c) && (byType === null || byType.includes(c)),
    );
    out = [...new Set([...out, ...added])];
  }

  return new Set(out);
}

/**
 * What this person may actually use.
 *
 * The practice's set, minus what their role never has, minus what their grant
 * does not cover.
 *
 * `membership` may be null — a caller whose membership row does not exist yet
 * is the pre-backfill case, and they get the practice's set unnarrowed. Same
 * rule as everywhere else: absence is not evidence of a restriction.
 */
export function effectiveCapabilities({ practice, membership, role = null }) {
  const held = capabilitiesOfPractice(practice);
  if (!membership) return held;

  const theirRole = membership.role ?? role;
  const excluded = ROLE_EXCLUDES[theirRole] ?? [];

  // An empty grant means the preset applies, which the model already resolves
  // on read. An empty array here would mean "may do nothing", and that is the
  // reading that locks somebody out of their own practice.
  const grant = membership.permissions?.length ? membership.permissions : null;

  return new Set(
    [...held].filter((c) => {
      if (excluded.includes(c)) return false;
      const needs = NEEDS_PERMISSION[c];
      if (!needs || !grant) return true;
      return grant.includes(needs);
    }),
  );
}

/** Convenience for a single check. */
export function can(capability, { practice, membership, role = null }) {
  return effectiveCapabilities({ practice, membership, role }).has(capability);
}

/**
 * The shape the clients read to build themselves.
 *
 * Sorted, so a response body that has not changed does not look as though it
 * has, and a diff between two practices is readable.
 */
export function describeCapabilities({ practice, membership, role = null }) {
  return {
    practiceType: practice?.practiceType ?? null,
    specialty: practice?.specialty ?? null,
    plan: practice?.plan ?? null,
    practice: [...capabilitiesOfPractice(practice)].sort(),
    effective: [...effectiveCapabilities({ practice, membership, role })].sort(),
  };
}

/**
 * Every capability, and what is holding back the ones this practice lacks.
 *
 * ---- Why anything needs this -------------------------------------------
 *
 * `DEPARTMENT` clears three independent gates: the practice type must be one
 * that has departments at all, the plan must pay for them, and the person must
 * hold MANAGE_DEPARTMENT. Miss any one and the app draws nothing — deliberately,
 * because a greyed section on a screen a solo doctor opens weekly is a
 * permanent advertisement for something that will never apply to them.
 *
 * That is right for the doctor and useless for the operator who has just set a
 * practice up and is looking at a screen with no departments on it. Nothing in
 * the console showed what a type and a plan added up to, so the only way to
 * find out was to read two tables in this file.
 *
 * Type first when both block it: a plan is a sale and a type is what the
 * organisation is, so "a clinic does not have departments" is the more useful
 * half of the answer and the one that does not resolve itself with money.
 */
export function explainCapabilities(practice) {
  const byType = ceiling(BY_TYPE, practice?.practiceType);
  const byPlan = ceiling(BY_PLAN, practice?.plan);
  const held = capabilitiesOfPractice(practice);

  return ALL.map((capability) => {
    const typeAllows = byType === null || byType.includes(capability);
    const planAllows = byPlan === null || byPlan.includes(capability);

    return {
      capability,
      has: held.has(capability),
      /// What stops it, or null when nothing does.
      blockedBy: held.has(capability) ? null : !typeAllows ? 'type' : !planAllows ? 'plan' : null,
      /**
       * And what a member still needs before they can use it.
       *
       * Sent even when the practice has the capability, because "the practice
       * has departments and this doctor cannot manage them" is the commonest
       * reason the section is missing from one person's screen and not
       * another's — and it is a conversation with the practice, not a sale.
       */
      needsPermission: NEEDS_PERMISSION[capability] ?? null,
    };
  });
}

export { ALL as ALL_CAPABILITIES, BY_TYPE, BY_PLAN, NEEDS_PERMISSION, ROLE_EXCLUDES };
